docs/library/espnow.rst

```diff
@@ -0,0 +1,917 @@
+          ``False``)
+
+        **ESP8266**: Keyword args may not be used on the ESP8266.
+
+        **Note:** The maximum number of peers which may be registered is 20
+        (`espnow.MAX_TOTAL_PEER_NUM`), with a maximum of 6
+        (`espnow.MAX_ENCRYPT_PEER_NUM`) of those peers with encryption enabled
+        (see `ESP_NOW_MAX_ENCRYPT_PEER_NUM <https://docs.espressif.com/
+        projects/esp-idf/en/latest/esp32/api-reference/network/
+        esp_now.html#c.ESP_NOW_MAX_ENCRYPT_PEER_NUM>`_ in the Espressif API
+        docs).
+
+    .. data:: Raises:
+
+        - ``OSError(num, "ESP_ERR_ESPNOW_NOT_INIT")`` if not initialised.
+        - ``OSError(num, "ESP_ERR_ESPNOW_EXIST")`` if ``mac`` is already
+          registered.
+        - ``OSError(num, "ESP_ERR_ESPNOW_FULL")`` if too many peers are
+          already registered.
+        - ``ValueError()`` on invalid keyword args or values.
+
+.. method:: ESPNow.del_peer(mac)
+
+    Deregister the peer associated with the provided ``mac`` address.
+
+    .. data:: Returns:
+
+        ``None``
+
+    .. data:: Raises:
+
+        - ``OSError(num, "ESP_ERR_ESPNOW_NOT_INIT")`` if not initialised.
+        - ``OSError(num, "ESP_ERR_ESPNOW_NOT_FOUND")`` if ``mac`` is not
+          registered.
+        - ``ValueError()`` on invalid ``mac`` values.
+
+.. method:: ESPNow.get_peer(mac) (ESP32 only)
+
+    Return information on a registered peer.
+
+    .. data:: Returns:
+
+        ``(mac, lmk, channel, ifidx, encrypt)``: a tuple of the "peer
+        info" associated with the ``mac`` address.
+
+    .. data:: Raises:
+
+        - ``OSError(num, "ESP_ERR_ESPNOW_NOT_INIT")`` if not initialised.
+        - ``OSError(num, "ESP_ERR_ESPNOW_NOT_FOUND")`` if ``mac`` is not
+          registered.
+        - ``ValueError()`` on invalid ``mac`` values.
+
+.. method:: ESPNow.peer_count() (ESP32 only)
+
+    Return the number of registered peers:
+
+    - ``(peer_num, encrypt_num)``: where
+
+      - ``peer_num`` is the number of peers which are registered, and
+      - ``encrypt_num`` is the number of encrypted peers.
+
+.. method:: ESPNow.get_peers() (ESP32 only)
+
+    Return the "peer info" parameters for all the registered peers (as a tuple
+    of tuples).
+
+.. method:: ESPNow.mod_peer(mac, lmk, [channel], [ifidx], [encrypt]) (ESP32 only)
+            ESPNow.mod_peer(mac, 'param'=value, ...) (ESP32 only)
+
+    Modify the parameters of the peer associated with the provided ``mac``
+    address. Parameters may be provided as positional or keyword arguments
+    (see `ESPNow.add_peer()`).
+
+Callback Methods
+----------------
+
+.. method:: ESPNow.irq(callback[, arg=None]) (ESP32 only)
+
+  Set a callback function to be called *as soon as possible* after a message has
+  been received from another ESPNow device. The function will be called with
+  ``arg`` as an argument, eg: ::
+
+          def recv_cb(e):
+              print(e.irecv(0))
+          e.irq(recv_cb, e)
```

Comment:
How about the callback always passes the `ESPNow` instance in as the first argument, and does not allow changing it?  That is, the signature is:
```
ESPNow.irq(handler)
```
And use is:
```py
def recv_cb(espnow):
    ...
espnow.irq(recv_cb)
```
That matches other things, eg `machine.Pin.irq(...)`.  It also simplifies the C code for espnow.

Reply:
I have changed the documentation (and example) to emphasise the use without the optiional *arg* argument.

However, it is necessary to be able to pass the `ESPNow` instance to the `ESPNowBase.irq()` method as passing the `ESPNowBase` singleton (from the `_espnow` C module) to the handler is not so useful as important parts of the user api are provided in rhe python module and would be confusing for users. Eg. it would not be possible to call `e.recv()` from the handler to read messages.

I *could* also remove the optional "arg" from `ESPNow.irq(callback. arg=None)` in `espnow.py`, but I would need to retain the "arg" in the _espnow C module. Let me know if you would prefer that.

Reply:
> I could also remove the optional "arg" from ESPNow.irq(callback. arg=None) in espnow.py, but I would need to retain the "arg" in the _espnow C module. Let me know if you would prefer that.

Yes please do that.

Reply:
Done.